### PR TITLE
Added test-tidy for dependent licenses.

### PR DESCRIPTION
**Pull request by asajeffrey**
Add a test-tidy lint for dependency licenses. Note that #12507 should land first.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because we don't test our lints

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12511)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @metajack @aneeshusa

File: python/tidy/servo_tidy/tidy.py [hunk 1]
```diff
@@ -642,6 +642,23 @@ def check_wpt_lint_errors(files):
...
+            filenames = progress_wrapper(filenames)
+        for filename in filenames:
+            if filename == "Cargo.toml":
+                filename = os.path.join(root, filename)
+                with open(filename, "r") as f:
+                    ok_licensed = False
+                    lines = f.read().splitlines(True)
+                    for idx, line in enumerate(lines):
+                        for license in licenses_dep_toml:
+                            ok_licensed |= (license in line)
```
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Not sure whether `license` would be a valid name. It's a built-in function :)

File: python/tidy/servo_tidy/tidy.py [hunk 2]
```diff
@@ -642,6 +642,23 @@ def check_wpt_lint_errors(files):
...
+def check_dep_license_errors(progress):
+    for root, directories, filenames in os.walk(".cargo"):
+        if progress:
+            filenames = progress_wrapper(filenames)
+        for filename in filenames:
+            if filename == "Cargo.toml":
+                filename = os.path.join(root, filename)
+                with open(filename, "r") as f:
+                    ok_licensed = False
+                    lines = f.read().splitlines(True)
```
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: could be `f.readlines()`

File: python/tidy/servo_tidy/tidy.py [hunk 3]
```diff
@@ -642,6 +642,23 @@ def check_wpt_lint_errors(files):
             yield ("WPT Lint Tool", "", "lint error(s) in Web Platform Tests: exit status {0}".format(returncode))
 
 
+def check_dep_license_errors(progress):
+    for root, directories, filenames in os.walk(".cargo"):
```
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: It'd be better if we print something like `'Checking licenses...'` before beginning the iterator.

File: python/tidy/servo_tidy/tidy.py [hunk 4]
```diff
@@ -675,8 +692,10 @@ def scan(only_changed_files=False, progress=True):
     errors = collect_errors_for_files(files_to_check, checking_functions, line_checking_functions)
     # wpt lint checks
     wpt_lint_errors = check_wpt_lint_errors(get_wpt_files(only_changed_files, progress))
+    # check dependecy licenses
+    dep_license_errors = check_dep_license_errors(progress)
```
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Also, doesn't this consume time? Wouldn't it be better if we check this when we run `./mach test-tidy --all` (i.e.,) `only_changed_files=False`?

File: python/tidy/servo_tidy/tidy.py (hunk 3 above)
Comment (asajeffrey) [inline python/tidy/servo_tidy/tidy.py]: Done.

File: python/tidy/servo_tidy/tidy.py (hunk 2 above)
Comment (asajeffrey) [inline python/tidy/servo_tidy/tidy.py]: Done.

File: python/tidy/servo_tidy/tidy.py (hunk 1 above)
Comment (asajeffrey) [inline python/tidy/servo_tidy/tidy.py]: Renamed to `license_line`.

File: python/tidy/servo_tidy/tidy.py (hunk 4 above)
Comment (asajeffrey) [inline python/tidy/servo_tidy/tidy.py]: Fixed.

**Comment by asajeffrey** [comment]
BTW, the progress indicator isn't printing correctly:
```
$ ./mach test-tidy --all
Checking files for tidiness...
  Progress: 100% (9297/9297)
Running the WPT lint...
  Progress: 100% (20057/20057)
Running the dependency licensing lint...
.cargo/git/checkouts/rust-clipboard-4a1584d52df56684/master/Cargo.toml:0: dependency should contain a valid license.
  Progress: 100% (1/1)2)6)
```
what is the magic I should be using?

File: python/tidy/servo_tidy/licenseck.py [hunk 5]
```diff
@@ -88,3 +88,40 @@
...
+    'license = "BSD-2-Clause"',
+    'license = "BSD-3-Clause"',
+    'license = "BSD-3-Clause/MIT"',
+    'license = "CC0-1.0"',
+    'license = "ISC"',
+    'license = "MIT / Apache-2.0"  # See https://github.com/rust-lang/cargo/issues/2174',
+    'license = "MIT / Apache-2.0"',
+    'license = "MIT OR Apache-2.0"',
+    'license = "MIT"',
+    'license = "MIT"',
```
Comment (aneeshusa) [inline python/tidy/servo_tidy/licenseck.py]: dupe?

**Comment by aneeshusa** [comment]
The `progress_wrapper` function needs an iterator as its argument, not a list like `filenames`, so that it can update the progress each time it yields an object. I'd recommend creating a helper function that `yield`s `Cargo.toml` files and thus returns an iterator, which can then be consumed by `progress_wrapper` if `progress` is `True`. (See `get_wpt_files` for an example usage.)

Also, it would be good to have the lint run even if progress is `False` :)

**Comment by Wafflespeanut** [comment]
But, the `progress_wrapper` converts the iterator into a list again. Let's make `progress_wrapper` take a list as input (and modify the related methods) instead?

File: python/tidy/servo_tidy/tidy.py [hunk 6]
```diff
@@ -642,6 +642,25 @@ def check_wpt_lint_errors(files):
             yield ("WPT Lint Tool", "", "lint error(s) in Web Platform Tests: exit status {0}".format(returncode))
 
 
+def check_dep_license_errors(only_changed_files=False, progress=True):
+    if not only_changed_files:
+        print '\nRunning the dependency licensing lint...'
+        for root, directories, filenames in os.walk(".cargo"):
+            if progress:
+                filenames = progress_wrapper(filenames)
```
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Yep, I missed this. As Aneesh said, this should be

``` python
filenames = progress_wrapper(filenames) if progress else filenames
```

**Comment by aneeshusa** [comment]
@Wafflespeanut I'd prefer to use iterators everywhere (for reduced memory usage), but your call. At the very least, I think we should still accept regular iterators as input to `progress_wrapper`.

**Comment by nox** [comment]
How come do we have to do something similar to grep when we could read the files as TOML files from Python?

File: python/tidy/servo_tidy/licenseck.py (hunk 5 above)
Comment (asajeffrey) [inline python/tidy/servo_tidy/licenseck.py]: Fixed.

File: python/tidy/servo_tidy/tidy.py (hunk 6 above)
Comment (asajeffrey) [inline python/tidy/servo_tidy/tidy.py]: Fixed. (Also fixed the indentation, bah humbug emacs python mode.)

**Comment by asajeffrey** [comment]
Yay, the progress meter now works!

**Comment by asajeffrey** [comment]
@nox: it would be a bit cleaner to parse the toml file, but would probably use a bit more memory unless there's a clever streaming toml parser out there.

**Comment by asajeffrey** [comment]
#12507 has landed, so we should be able to land this one. @larsbergstrom?

**Comment by larsbergstrom** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12511)*

**Comment by larsbergstrom** [comment]
looks good! squash and r=me

**Comment by asajeffrey** [comment]
Squashed. @bors-servo r=larsbergstrom

**Pull request merged**